base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -0,0 +1,259 @@
+    types::HashOutput,
+};
+use log::*;
+use std::sync::{Arc, RwLock, RwLockReadGuard};
+
+const LOG_TARGET: &str = "core::chain_storage::database";
+
+pub enum MmrTree {
+    Utxo,
+    Kernel,
+    RangeProof,
+    Header,
+}
+
+pub trait BlockchainBackend: Send + Sync {
+    fn write(&self, tx: DbTransaction) -> Result<(), ChainStorageError>;
+    fn get(&self, key: &DbKey) -> Result<Option<DbValue>, ChainStorageError>;
+    fn contains(&self, key: &DbKey) -> Result<bool, ChainStorageError>;
+    fn get_mmr_root(&self, tree: MmrTree) -> Result<HashOutput, ChainStorageError>;
+}
+
+// Private macro that pulls out all the boiler plate of extracting a DB query result from its variants
+macro_rules! fetch {
+    ($self:ident, $key_val:expr, $key_var:ident) => {{
+        let key = DbKey::$key_var($key_val);
+        match $self.db.get(&key) {
+            Ok(None) => Ok(None),
+            Ok(Some(DbValue::$key_var(k))) => Ok(Some(*k)),
+            Ok(Some(other)) => unexpected_result(key, other),
+            Err(e) => log_error(key, e),
+        }
+    }};
+}
+
+pub struct BlockChainDatabase<T>
+where T: BlockchainBackend
+{
+    metadata: Arc<RwLock<ChainMetadata>>,
+    db: Arc<T>,
+}
+
+impl<T> BlockChainDatabase<T>
+where T: BlockchainBackend
+{
+    /// Reads the blockchain metadata (block height etc) from the underlying backend and returns it.
+    fn read_metadata(db: &T) -> Result<ChainMetadata, ChainStorageError> {
+        let height = match db.get(&DbKey::Metadata(MetadataKey::ChainHeight)) {
+            Ok(Some(DbValue::Metadata(MetadataValue::ChainHeight(v)))) => v,
+            Ok(None) => {
+                warn!(
+                    target: LOG_TARGET,
+                    "The chain height entry is not present in the database. Assuming the database is empty."
+                );
+                0
+            },
+            Err(e) => return log_error(DbKey::Metadata(MetadataKey::ChainHeight), e),
+            Ok(Some(other)) => return unexpected_result(DbKey::Metadata(MetadataKey::ChainHeight), other),
+        };
+
+        let work = match db.get(&DbKey::Metadata(MetadataKey::AccumulatedWork)) {
+            Ok(Some(DbValue::Metadata(MetadataValue::AccumulatedWork(v)))) => v,
+            Ok(None) => {
+                warn!(
+                    target: LOG_TARGET,
+                    "The accumulated work entry is not present in the database. Assuming the database is empty."
+                );
+                0
+            },
+            Err(e) => return log_error(DbKey::Metadata(MetadataKey::AccumulatedWork), e),
+            Ok(Some(other)) => return unexpected_result(DbKey::Metadata(MetadataKey::AccumulatedWork), other),
+        };
+
+        Ok(ChainMetadata {
+            height_of_longest_chain: height,
+            greatest_accumulated_work: work,
+        })
+    }
+
+    /// Creates a new `BlockchainDatabase` using the provided backend.
+    pub fn new(db: T) -> Result<Self, ChainStorageError> {
```

Comment:
Nit: Personally I like having the constructors being the first method in the impl.

Reply:
Me too.